File: oxidize-golang/core/model/offload.go
package model

import (
	"errors"
	"math"
	"sort"
)

// OffloadPolicy controls how a model is partitioned between CPU and GPU.
type OffloadPolicy int

// Recognised policies.
const (
	OffloadPolicyLayerMajor OffloadPolicy = iota
	OffloadPolicyRowMajor
	OffloadPolicyPipeline
)

// String returns a human-readable label.
func (p OffloadPolicy) String() string {
	switch p {
	case OffloadPolicyLayerMajor:
		return "layer-major"
	case OffloadPolicyRowMajor:
		return "row-major"
	case OffloadPolicyPipeline:
		return "pipeline"
	}
	return "unknown"
}

// DeviceMemory represents a single device's memory budget.
type DeviceMemory struct {
	DeviceID int
	Backend  string
	Bytes    int64
}

// LayerOffloadPlan mirrors LayerOffloadPlan.
type LayerOffloadPlan struct {
	Layers []LayerAssignment
}

// LayerAssignment describes where a single layer lives.
type LayerAssignment struct {
	LayerIndex int
	DeviceID   int
	Backend    string
	Bytes      int64
}

// TotalBytes returns the sum of bytes across all layer assignments.
func (p *LayerOffloadPlan) TotalBytes() int64 {
	var total int64
	for _, l := range p.Layers {
		total += l.Bytes
	}
	return total
}

// ByDevice groups assignments by device id.
func (p *LayerOffloadPlan) ByDevice() map[int]int64 {
	out := map[int]int64{}
	for _, l := range p.Layers {
		out[l.DeviceID] += l.Bytes
	}
	return out
}

// MultiGpuOffloadPlan mirrors MultiGpuOffloadPlan.
type MultiGpuOffloadPlan struct {
	Stages    []PipelineStage
	Policy    OffloadPolicy
	TotalBytes int64
}

// PipelineStage mirrors PipelineStage.
type PipelineStage struct {
	StageID     int
	LayerRange  [2]int
	DeviceID    int
	Backend     string
	Bytes       int64
	MicroBatch  int
}

// NewMultiGpuOffloadPlan creates an empty plan with the given policy.
func NewMultiGpuOffloadPlan(policy OffloadPolicy) *MultiGpuOffloadPlan {
	return &MultiGpuOffloadPlan{Policy: policy}
}

// AddStage appends a stage to the plan.
func (p *MultiGpuOffloadPlan) AddStage(stage PipelineStage) {
	if stage.MicroBatch <= 0 {
		stage.MicroBatch = 1
	}
	p.Stages = append(p.Stages, stage)
	p.TotalBytes += stage.Bytes
}

// Validate ensures the plan covers all layers without gaps.
func (p *MultiGpuOffloadPlan) Validate(totalLayers int) error {
	if len(p.Stages) == 0 {
		return errors.New("offload: no stages")
	}
	sort.Slice(p.Stages, func(i, j int) bool { return p.Stages[i].LayerRange[0] < p.Stages[j].LayerRange[0] })
	for i, s := range p.Stages {
		if s.LayerRange[0] < 0 || s.LayerRange[1] > totalLayers {
			return errors.New("offload: layer range out of bounds")
		}
		if s.LayerRange[0] >= s.LayerRange[1] {
			return errors.New("offload: empty layer range")
		}
		if i > 0 && s.LayerRange[0] != p.Stages[i-1].LayerRange[1] {
			return errors.New("offload: gap or overlap between stages")
		}
	}
	return nil
}

// LayerOffloadPlanner mirrors LayerOffloadPlanner.
type LayerOffloadPlanner struct {
	Devices []DeviceMemory
	Policy  OffloadPolicy
}

// NewLayerOffloadPlanner constructs a planner.
func NewLayerOffloadPlanner(devices []DeviceMemory, policy OffloadPolicy) *LayerOffloadPlanner {
	return &LayerOffloadPlanner{Devices: devices, Policy: policy}
}

// Plan distributes `layerCount` layers across the available devices.
func (p *LayerOffloadPlanner) Plan(layerCount int, bytesPerLayer int64) *LayerOffloadPlan {
	if len(p.Devices) == 0 || layerCount == 0 {
		return &LayerOffloadPlan{}
	}
	plan := &LayerOffloadPlan{}
	switch p.Policy {
	case OffloadPolicyLayerMajor:
		per := layerCount / len(p.Devices)
		rem := layerCount % len(p.Devices)
		idx := 0
		for d := range p.Devices {
			end := idx + per
			if d < rem {
				end++
			}
			for i := idx; i < end; i++ {
				plan.Layers = append(plan.Layers, LayerAssignment{
					LayerIndex: i,
					DeviceID:   p.Devices[d].DeviceID,
					Backend:    p.Devices[d].Backend,
					Bytes:      bytesPerLayer,
				})
			}
			idx = end
		}
	case OffloadPolicyRowMajor, OffloadPolicyPipeline:
		// Treat as a contiguous block
		for i := 0; i < layerCount; i++ {
			d := p.Devices[i%len(p.Devices)]
			plan.Layers = append(plan.Layers, LayerAssignment{
				LayerIndex: i,
				DeviceID:   d.DeviceID,
				Backend:    d.Backend,
				Bytes:      bytesPerLayer,
			})
		}
	}
	return plan
}

// GpuOffloadConfig mirrors GpuOffloadConfig.
type GpuOffloadConfig struct {
	Enabled        bool
	MaxLayersOnGpu int
	FallbackToCPU  bool
	UseAsyncCopy   bool
}

// OffloadMetrics tracks offload statistics.
type OffloadMetrics struct {
	TotalTransfers int64
	BytesTransferred int64
	AvgLatencyMicros float32
}

// AvgBytesPerTransfer returns the average bytes per transfer.
func (m OffloadMetrics) AvgBytesPerTransfer() float32 {
	if m.TotalTransfers == 0 {
		return 0
	}
	return float32(m.BytesTransferred / m.TotalTransfers)
}

// silence unused import
var _ = math.Ceil

File: oxidize-python/oxidize_python/core/model/offload.py
"""Multi-GPU offload planning mirroring oxidize-golang/core/model/offload.go."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum


class OffloadPolicy(IntEnum):
    LAYER_MAJOR = 0
    ROW_MAJOR = 1
    PIPELINE = 2

    def __str__(self) -> str:
        if self == OffloadPolicy.LAYER_MAJOR:
            return "layer-major"
        if self == OffloadPolicy.ROW_MAJOR:
            return "row-major"
        if self == OffloadPolicy.PIPELINE:
            return "pipeline"
        return "unknown"


@dataclass
class DeviceMemory:
    device_id: int
    backend: str
    bytes: int


@dataclass
class LayerAssignment:
    layer_index: int
    device_id: int
    backend: str
    bytes: int


@dataclass
class LayerOffloadPlan:
    layers: list[LayerAssignment] = field(default_factory=list)

    def total_bytes(self) -> int:
        return sum(layer.bytes for layer in self.layers)

    def by_device(self) -> dict[int, int]:
        out: dict[int, int] = {}
        for layer in self.layers:
            out[layer.device_id] = out.get(layer.device_id, 0) + layer.bytes
        return out


@dataclass
class PipelineStage:
    stage_id: int
    layer_range: tuple[int, int]
    device_id: int
    backend: str
    bytes: int
    micro_batch: int = 1


@dataclass
class MultiGpuOffloadPlan:
    stages: list[PipelineStage] = field(default_factory=list)
    policy: OffloadPolicy = OffloadPolicy.LAYER_MAJOR
    total_bytes: int = 0

    def add_stage(self, stage: PipelineStage) -> None:
        if stage.micro_batch <= 0:
            stage.micro_batch = 1
        self.stages.append(stage)
        self.total_bytes += stage.bytes

    def validate(self, total_layers: int) -> None:
        if not self.stages:
            raise OffloadError("no stages")
        sorted_stages = sorted(self.stages, key=lambda s: s.layer_range[0])
        for i, s in enumerate(sorted_stages):
            lo, hi = s.layer_range
            if lo < 0 or hi > total_layers:
                raise OffloadError("layer range out of bounds")
            if lo >= hi:
                raise OffloadError("empty layer range")
            if i > 0 and lo != sorted_stages[i - 1].layer_range[1]:
                raise OffloadError("gap or overlap between stages")


class OffloadError(Exception):
    pass


class LayerOffloadPlanner:
    def __init__(self, devices: list[DeviceMemory], policy: OffloadPolicy) -> None:
        self.devices = devices
        self.policy = policy

    def plan(self, layer_count: int, bytes_per_layer: int) -> LayerOffloadPlan:
        if not self.devices or layer_count == 0:
            return LayerOffloadPlan()
        plan = LayerOffloadPlan()
        if self.policy == OffloadPolicy.LAYER_MAJOR:
            per = layer_count // len(self.devices)
            rem = layer_count % len(self.devices)
            idx = 0
            for d, dev in enumerate(self.devices):
                end = idx + per + (1 if d < rem else 0)
                for i in range(idx, end):
                    plan.layers.append(
                        LayerAssignment(i, dev.device_id, dev.backend, bytes_per_layer)
                    )
                idx = end
        else:
            for i in range(layer_count):
                dev = self.devices[i % len(self.devices)]
                plan.layers.append(
                    LayerAssignment(i, dev.device_id, dev.backend, bytes_per_layer)
                )
        return plan


@dataclass
class GpuOffloadConfig:
    enabled: bool = False
    max_layers_on_gpu: int = 0
    fallback_to_cpu: bool = True
    use_async_copy: bool = False


@dataclass
class OffloadMetrics:
    total_transfers: int = 0
    bytes_transferred: int = 0
    avg_latency_micros: float = 0.0

    def avg_bytes_per_transfer(self) -> float:
        if self.total_transfers == 0:
            return 0.0
        return self.bytes_transferred / self.total_transfers
